pkg/agent/controller/bgp/controller.go

```diff
@@ -991,3 +983,94 @@ func (c *Controller) GetBGPPeerStatus(ctx context.Context) ([]bgp.PeerStatus, er
 	}
 	return peers, nil
 }
+
+// GetBGPRoutes returns the advertised/received routes to/from the given peerAddress.
+// If given peerAddress is empty, routes from all BGP Peers are returned.
+func (c *Controller) GetBGPRoutes(ctx context.Context, routeType bgp.RouteType, peerAddress string) (sets.Set[string], error) {
+	bgpPeerAddresses := c.getBGPPeerAddresses()
+	if bgpPeerAddresses == nil {
+		return nil, ErrBGPPolicyNotFound
+	}
+
+	// Validate peerAddress, consider empty peerAddress as valid.
+	if peerAddress != "" {
+		if !utilnet.IsIPv6String(peerAddress) && !utilnet.IsIPv4String(peerAddress) {
+			return nil, fmt.Errorf("invalid peer address: %s", peerAddress)
+		}
+		if !slices.Contains(bgpPeerAddresses, peerAddress) {
+			return nil, fmt.Errorf("bgp peer %s not found", peerAddress)
+		}
+		bgpPeerAddresses = []string{peerAddress}
+	}
+
+	bgpRoutes := make(sets.Set[string])
+	if routeType == bgp.RouteAdvertised {
+		// Advertised routes are same for all bgp peers.
+		bgpRoutesAdvertised := c.getBgpRoutesAdvertised()
+		if bgpRoutesAdvertised == nil {
+			return nil, ErrBGPPolicyNotFound
+		}
+		for route := range bgpRoutesAdvertised {
+			bgpRoutes.Insert(route.Prefix)
+		}
+	} else {
+		bgpRoutesReceived, err := c.getBgpRoutesReceived(ctx, bgpPeerAddresses)
+		if err != nil {
+			return nil, err
+		}
+		for route := range bgpRoutesReceived {
+			bgpRoutes.Insert(route.Prefix)
+		}
+	}
+	return bgpRoutes, nil
+}
+
+func (c *Controller) getBgpServer() bgp.Interface {
+	c.bgpPolicyStateMutex.RLock()
+	defer c.bgpPolicyStateMutex.RUnlock()
+
+	if c.bgpPolicyState == nil {
+		return nil
+	}
+	return c.bgpPolicyState.bgpServer
+}
+
+func (c *Controller) getBgpRoutesAdvertised() sets.Set[bgp.Route] {
+	c.bgpPolicyStateMutex.RLock()
+	defer c.bgpPolicyStateMutex.RUnlock()
+
+	if c.bgpPolicyState == nil {
+		return nil
+	}
+	return c.bgpPolicyState.routes
+}
+
+func (c *Controller) getBgpRoutesReceived(ctx context.Context, bgpPeerAddresses []string) (sets.Set[bgp.Route], error) {
+	bgpServer := c.getBgpServer()
+	if bgpServer == nil {
+		return nil, ErrBGPPolicyNotFound
+	}
+	bgpRoutesReceived := make(sets.Set[bgp.Route])
+	for _, bgpPeerAddress := range bgpPeerAddresses {
+		routesReceived, err := bgpServer.GetRoutes(ctx, bgp.RouteReceived, bgpPeerAddress)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get bgp routes: %w", err)
+		}
+		bgpRoutesReceived.Insert(routesReceived...)
+	}
+	return bgpRoutesReceived, nil
+}
+
+func (c *Controller) getBGPPeerAddresses() []string {
```

Comment:
How about using a string set here? Then you don't need to use `slice.Contain` in `GetBGPRoutes`, which should be less efficient.